### ISSUE TITLE
falcon + gunicorn + gevent SSL monkey_patch crash loop

### ISSUE THREAD

**Issue by guongle-ssense** [DataDog/dd-trace-py#1624]
### Which version of dd-trace-py are you using?
0.41.0
### Which version of the libraries are you using?

attrs==19.3.0
boto3==1.12.39
botocore==1.15.39
falcon==2.0.0
flake8==3.7.9
gevent==1.4.0
gunicorn==20.0.4
jsonschema==3.2.0
more-itertools==8.0.2
pyflakes==2.1.1
pyrsistent==0.15.6
python-dateutil==2.8.1
python-http-client==3.2.4
s3transfer==0.3.3
urllib3==1.25.8
zipp==0.6.0
datadog==0.35.0
openapi-core==0.13.3
dependency-injector==3.15.6
ddtrace==0.41.0

### How can we reproduce your problem?
a simple falcon `api.py` file

```
import falcon

from src.components.controllers.healtcheck_controller import HealthCheckController
from src.components.enums.routes import Routes
from configs.general import APP_NAME

# Disable test coverage for this file
__test__ = False

app = falcon.API()

# Load healthcheck routes
health_check_controller = HealthCheckController()
app.add_route(str(Routes.HEALTH_CHECK.value), health_check_controller)
```

Then the command is:

```
ddtrace-run gunicorn src.api.api:app -b 0.0.0.0:80 --workers=3 --reload --worker-class gevent
```
### What is the result that you get?
At first I would get this:
```bash
/usr/local/lib/python3.7/site-packages/gunicorn/workers/ggevent.py:53: MonkeyPatchWarning: Monkey-patching ssl after ssl has already been imported may lead to errors, including RecursionError on Python 3.6. It may also silently lead to incorrect behaviour on Python 3.7. Please monkey-patch earlier. See https://github.com/gevent/gevent/issues/1016. Modules that had direct imports (NOT patched): ['botocore.httpsession (/usr/local/lib/python3.7/site-packages/botocore/httpsession.py)', 'urllib3.util (/usr/local/lib/python3.7/site-packages/urllib3/util/__init__.py)', 'urllib3.util.ssl_ (/usr/local/lib/python3.7/site-packages/urllib3/util/ssl_.py)'].
  monkey.patch_all()
````

Then as per:
https://github.com/DataDog/dd-trace-py/issues/506

The sequence is now:
```
import gevent.monkey
gevent.monkey.patch_all()
from requests.packages.urllib3.util.ssl_ import create_urllib3_context
create_urllib3_context()

import falcon
from ddtrace import tracer, patch
patch(falcon=True)
```

I have a new crash loop issue:
```bash
  File "/code/src/api/api.py", line 4, in <module>
    create_urllib3_context()
  File "/usr/local/lib/python3.7/site-packages/urllib3/util/ssl_.py", line 273, in create_urllib3_context
    context.options |= options
  File "/usr/local/lib/python3.7/ssl.py", line 518, in options
    super(SSLContext, SSLContext).options.__set__(self, value)
  File "/usr/local/lib/python3.7/ssl.py", line 518, in options
    super(SSLContext, SSLContext).options.__set__(self, value)
  File "/usr/local/lib/python3.7/ssl.py", line 518, in options
    super(SSLContext, SSLContext).options.__set__(self, value)
  [Previous line repeated 485 more times]
RecursionError: maximum recursion depth exceeded while calling a Python object
```

I have also tried the --preload directive:
```bash
ddtrace-run gunicorn src.api.api:app -b 0.0.0.0:80 --workers=3 --preload --reload --worker-class gevent
- DATADOG TRACER DIAGNOSTIC - Agent not reachable. Exception raised: [Errno 111] Connection refused
- DATADOG TRACER DIAGNOSTIC - Agent not reachable. Exception raised: [Errno 111] Connection refused
/code/src/api/api.py:2: MonkeyPatchWarning: Monkey-patching ssl after ssl has already been imported may lead to errors, including RecursionError on Python 3.6. It may also silently lead to incorrect behaviour on Python 3.7. Please monkey-patch earlier. See https://github.com/gevent/gevent/issues/1016. Modules that had direct imports (NOT patched): ['botocore.httpsession (/usr/local/lib/python3.7/site-packages/botocore/httpsession.py)', 'urllib3.util (/usr/local/lib/python3.7/site-packages/urllib3/util/__init__.py)', 'urllib3.util.ssl_ (/usr/local/lib/python3.7/site-packages/urllib3/util/ssl_.py)'].
  gevent.monkey.patch_all()

Error: maximum recursion depth exceeded while calling a Python object
```
### What is the result that you expected?
for my app to be up without the crash loop

Anyone faced this in falcon? I've seen the Django and other python solutions and followed those, so far the common theme is to monkey_patch all first however I have tried that but the crash still happens.

**Comment by Kyle-Verhoog** [comment]
Hi @guongle-ssense! Thanks for the report and the repro!

I put up a fix in #1629 which solved the issue for me with your example. Can you confirm it as well? It can be installed with

```
pip install git+https://github.com/datadog/dd-trace-py.git@kyle-verhoog/gevent
```

**Comment by Kyle-Verhoog** [comment]
For any future readers we have since also added an environment variable when using `ddtrace-run`, [`DD_GEVENT_PATCH_ALL`](https://ddtrace.readthedocs.io/en/stable/integrations.html#gevent) which will do gevent's monkey patching before ddtrace is loaded which should help deal with conflicts like the one described here.

Going to close the issue now. Please feel free to re-open if the issue is still occurring!
